### ISSUE TITLE
When all parameters are unused by forward pass in a process, backward will not work with DDP.

### ISSUE THREAD

**Issue by sinannasir** [pytorch/pytorch#40566]
We realized that when the `forward` pass produces empty `list(_find_tensors(output))` - referring the code below - in some processes with `find_unused_parameters = True`, the backward() will not work for DDP.

https://github.com/pytorch/pytorch/blob/e490352dc4ad4f61c795bfdae968954c7d916861/torch/nn/parallel/distributed.py#L521-L522

With today's reducer it's impossible to make all parameters unused, as we would require at least one `autograd_hook` invocation to process unused parameters. The loss function may take an empty list and then produce some irrelevant tensor. In that case applications can still call backward, but they cannot run it with today's DDP.

cc: @mrshenli @pritamdamania87 @zhaojuanmao 
This issue was realized in pull request: #40407
